### ISSUE TITLE
Add support for tracing streaming response bodies 

### ISSUE THREAD

**Issue by pcriv** [DataDog/dd-trace-rb#1934]
Currently using rack and streaming response bodies, Datadog will only show a trace until the server replies with a status code while the body is still being sent. This creates an inaccurate latency metric.

**Comment by delner** [comment]
Yeah this is an interesting problem, from a trace model perspective. What would you expect a trace to look like? Or the latency metric to measure?

If we can have a clearer picture of current behavior vs expected behavior, we can more easily assess and plan this.

**Comment by pcriv** [comment]
Ideally, I would like to not finish the trace until the last byte has been delivered. The latency would measure the whole length of the request which in our case, downloads, could take sometimes even hours for slow clients.

**Comment by delner** [comment]
I see. It's tricky because we need a mounting point in Rack for instrumentation to attach to, presumably with high precedence than this streaming behavior.

Right now, our instrumentation is just middleware: if its placed high enough in the stack, and the request doesn't spin off any asynchronous behavior, it should trace it.

 - If middleware completes before the request does, it doesn't capture preceding behavior.
 - If the request spins off an asynchronous callback executed in a different thread, the execution context of the callback is different than where the request originated, thus the callback span isn't automatically associated with the request span to form a contiguous trace.

Can you describe your Rack stack a bit more? I think we're going to need more technical detail on why the instrumentation isn't capturing this to assess what we can do about it.

**Comment by pcriv** [comment]
We use a gem that was built in-house at WeTransfer. It's called [fast_send](https://github.com/WeTransfer/fast_send)
It's a Rack middleware to send large, long-running Rack responses via file buffers, it uses Rack hijacking so it takes the usual Rack handling out of the response writing. 
In detail, the middleware takes Ruby Filehandles, one by one, and sends the file to the Rack webserver socket, calling non-blocking `sendfile(2)`  on Linux for us.

**Comment by pcriv** [comment]
I just saw that we did our own integration for fixing this for Appsignal, which we use now and we are trying to replace it with Datadog: https://github.com/WeTransfer/appsignal_extensions/blob/master/lib/appsignal_extensions/middleware.rb

**Comment by Strech** [comment]
Hey @pcriv sorry, but the link leads to private repository

**Comment by ivoanjo** [comment]
@Strech I'm guessing @pcriv is closing this since the issue was opened a long time ago ;)

Thanks Pablo for the cleanup and sorry we never got around to this one >_>
